### PR TITLE
Add Flattener

### PR DESCRIPTION
**Pull request by Jozwiaczek**
resolves: #156

### CODE REVIEW

File: waffle-cli/src/cli.ts [hunk 1]
```diff
@@ -1,8 +1,12 @@
 import {compileProject} from '@ethereum-waffle/compiler';
+import {flattenProject} from '@ethereum-waffle/compiler/src/flattener';
```
Comment (msieczko) [inline waffle-cli/src/cli.ts:2]: Import should look like this:
```typescript
import {compileProject, flattenProject} from '@ethereum-waffle/compiler';
```

Add an export to `waffle-compiler/src/index.ts` to achieve this.

File: waffle-cli/src/cli.ts [hunk 2]
```diff
@@ -30,7 +37,7 @@ function exitWithError(error: string) {
 const USAGE = `
   Usage:
 
-    waffle [config-file] [options]
+    waffle [compile|flatten] [config-file] [options]
 
     Compiles solidity source code
```
Comment (msieczko) [inline waffle-cli/src/cli.ts:43]: I'd add some description of the new commands usage.

File: waffle-cli/src/cli.ts [hunk 3]
```diff
@@ -13,10 +17,13 @@ export async function runCli(args: string[]) {
...
-    if (other.length > 1) {
+    if (other.length > 2) {
       exitWithError('Error: Too many arguments!');
     } else {
-      return compileProject(args[0]);
+      if (isTypeFlatten) {
+        return flattenProject(configFile);
+      }
+      return compileProject(configFile);
     }
```
Comment (msieczko) [inline waffle-cli/src/cli.ts:27]: Could be extracted to a new function `handleCommands(...)`.

File: waffle-cli/src/cli.ts [hunk 4]
```diff
@@ -1,8 +1,12 @@
...
 
 export async function runCli(args: string[]) {
   const options = args.filter((x) => x.startsWith('-'));
   const other = args.filter((x) => !x.startsWith('-'));
+  const isTypeFlatten = args[0] === ('flatten');
+  const isTypeProvided = isTypeFlatten || args[0] === 'compile';
+  const configFile = isTypeProvided ? args[1] : args[0];
 
   if (options.length > 0) {
     for (const option of options) {
```
Comment (msieczko) [inline waffle-cli/src/cli.ts:12]: Could be extracted to a new function `handleOptions(...)`.

File: waffle-compiler/src/config/config.ts [hunk 5]
```diff
@@ -2,7 +2,9 @@ export interface Config {
   /** Location of the project sources e.g. ./contracts */
   sourceDirectory: string;
   /** Location of the compilation artifacts e.g. ./dist */
-  outputDirectory: string;
+  compileOutputDirectory: string;
+  /** Location of the compilation artifacts e.g. ./dist */
+  flattenOutputDirectory: string;
```
Comment (msieczko) [inline waffle-compiler/src/config/config.ts:7]: Add outputDirectory to `LegacyConfig` type.

File: waffle-compiler/src/config/inputToConfig.ts [hunk 6]
```diff
@@ -16,8 +16,8 @@ export function inputToConfig(input: InputConfig) {
         deprecate('sourcesPath', 'Use sourceDirectory instead.');
         set('sourceDirectory', input[key]);
       } else if (key === 'targetPath') {
-        deprecate('targetPath', 'Use outputDirectory instead.');
-        set('outputDirectory', input[key]);
+        deprecate('targetPath', 'Use compileOutputDirectory instead.');
+        set('compileOutputDirectory', input[key]);
```
Comment (msieczko) [inline waffle-compiler/src/config/inputToConfig.ts:20]: You should also handle `outputDirectory` configuration option that you are deprecating in this PR.

File: waffle-compiler/src/config/inputToConfig.ts [hunk 7]
```diff
@@ -48,7 +48,8 @@ export function inputToConfig(input: InputConfig) {
 function validate(config: any): asserts config is Config {
   const success = (
     checkSourceDirectory(config.sourceDirectory) &&
-    checkOutputDirectory(config.outputDirectory) &&
+    checkCompileOutputDirectory(config.compileOutputDirectory) &&
+    checkFlattenOutputDirectory(config.flattenOutputDirectory) &&
     checkNodeModulesDirectory(config.nodeModulesDirectory) &&
     checkCompilerType(config.compilerType) &&
     checkCompilerVersion(config.compilerVersion) &&
```
Comment (msieczko) [inline waffle-compiler/src/config/inputToConfig.ts:55]: Not related to the task at hand but looks like a good place to use [@restless/sanitizers](https://www.npmjs.com/package/@restless/sanitizers).

File: waffle-cli/README.md [hunk 8]
```diff
@@ -155,7 +155,7 @@ Example configuration file looks like this (all fields optional):
 ```json
 {
   "sourceDirectory": "./custom_contracts",
-  "outputDirectory": "./custom_build",
+  "compileOutputDirectory": "./custom_build",
```
Comment (msieczko) [inline waffle-cli/README.md:158]: I'd describe the new feature in `README.md` file as well.

File: waffle-compiler/src/flattener.ts [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    });
+  });
+}
+
+async function processOutput(output: any, config: Config) {
+  if (output.errors) {
+    console.error(formatErrors(output.errors));
+  }
+  if (anyNonWarningErrors(output.errors)) {
+    throw new Error('Flattening failed');
```
Comment (msieczko) [inline waffle-compiler/src/flattener.ts:58]: Specify type instead of using any. If you did you would immediately notice that `output` will never have `errors` property 😀

File: waffle-compiler/src/flattener.ts [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+  return typeof error === 'string' ? error : error.formattedMessage;
+}
+
+const fsOps = {
+  createDirectory: mkdirp.sync,
+  writeFile: fsx.writeFileSync
+};
+
+function saveToFile(
+  output: any,
```
Comment (msieczko) [inline waffle-compiler/src/flattener.ts:82]: Specify type

File: waffle-compiler/test/flattener/testSource/cycle1.sol [hunk 11]
```diff
@@ -0,0 +1,5 @@
+import "./cycle2.sol";
```
Comment (msieczko) [inline waffle-compiler/test/flattener/testSource/cycle1.sol:1]: Every solidity file needs to have a `pragma` statement. I'd add them to these test files to make the test more realistic.

File: waffle-compiler/test/flattener/flattener.ts [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    await fs.remove(flattenOutputDirectory);
+  });
+
+  it('should include the parent if the child file', async () => {
+    await flattenAndSave(config);
+    const file = fs.readFileSync('./test/flattener/flattenFiles/child.sol', 'utf8');
+    expect(file).includes('contract Parent');
+  });
+
+  it('should leave multiple solidity pragmas', async () => {
```
Comment (msieczko) [inline waffle-compiler/test/flattener/flattener.ts:26]: The output file of a flattener generally should **not** include multiple solidity pragmas. This tests passes because they are left in `// Dependency file: ...` comments.

I'd leave the original pragma of each dependency file commented out in the flattener output file like this: 
```solidity
// Dependency file: openzeppelin-solidity/contracts/access/roles/PauserRole.sol
// pragma solidity ^0.5.0;

contract PauserRole {
    ...
}
```
...and then change the name and expect statement of this test to reflect that we are checking commented out pragmas.

File: waffle-compiler/test/flattener/flattener.ts [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+  flattenOutputDirectory,
+  compilerVersion
+});
+
+describe('flattening', () => {
+  after(async () => {
+    await fs.remove(flattenOutputDirectory);
+  });
+
+  it('should include the parent if the child file', async () => {
```
Comment (msieczko) [inline waffle-compiler/test/flattener/flattener.ts:20]:
```suggestion
  it('should include the parent in the child file', async () => {
```

File: waffle-compiler/src/flattener.ts [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    const fileName = path.parse(contract[0].url).base;
+    const filePath = join(outputDirectory, fileName);
+    let dependencies = '';
+    contract.map((dependency) => {
+      if (dependency !== contract[0]) {
+        dependencies += '\n\n' + `// Dependency file: ${dependency.url}`;
+      }
+      dependencies += dependency.source;
+    });
+    fileSystem.writeFile(filePath, dependencies);
```
Comment (msieczko) [inline waffle-compiler/src/flattener.ts:98]: If you look at an example [flattened contract](https://etherscan.io/address/0x6b175474e89094c44da98b954eedeac495271d0f#code), you'll see that dependencies above the flattened contract. I'd do the same here.

File: waffle-compiler/src/flattener.ts [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+    return gatherSources(
+      [contract],
+      '.',
+      resolver
+    );
+  }));
+
+  return contractsDependency.map((contract: Array<GatheredContractInterface>) => {
+    return contract.map((dependency) => {
+      dependency.source = dependency.source.replace(IMPORT_SOLIDITY_REGEX, '');
```
Comment (msieczko) [inline waffle-compiler/src/flattener.ts:47]: The output file has many empty lines. Maybe consider deleting empty lines resulting from removing imports? 🤔

File: waffle-compiler/src/flattener.ts [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+      return dependency;
+    });
+  });
```
Comment (msieczko) [inline waffle-compiler/src/flattener.ts:50]: `contractsDependency` does not reflect the meaning of the variable. It's an array of contracts with all of their dependency files. Instead of trying to come up with a good name for it, I would combine these two iterations into one 😀

File: waffle-compiler/test/flattener/flattener.ts (hunk 13 above)
Comment (msieczko) [reply waffle-compiler/test/flattener/flattener.ts:20]: Also, this test doesn't check much. Why not create the expected file by hand and compare whole file contents?

File: waffle-compiler/src/flattener.ts (hunk 15 above)
Comment (Jozwiaczek) [reply waffle-compiler/src/flattener.ts:47]: change regex to removing empty lines after imports

File: waffle-compiler/src/config/inputToConfig.ts [hunk 17]
```diff
@@ -16,8 +16,8 @@ export function inputToConfig(input: InputConfig) {
         deprecate('sourcesPath', 'Use sourceDirectory instead.');
         set('sourceDirectory', input[key]);
       } else if (key === 'targetPath') {
-        deprecate('targetPath', 'Use outputDirectory instead.');
-        set('outputDirectory', input[key]);
+        deprecate('targetPath', 'Use compileOutputDirectory instead.');
```
Comment (sz-piotr) [inline waffle-compiler/src/config/inputToConfig.ts:19]: This is not that easy. You cannot just rename the configuration variable like this. Consider what will happen if this gets released.

Somebody had this nice config:
```js
// .waffle.config.json
{
  "outputDirectory": "build"
}
```
This PR changes the configuration to not use `outputDirectory` but `compileOutputDirectory`. Yet not only is this person not notified about this change, his previous configuration stops working.

There is no need to introduce this unnecessary breaking change.

Comment (sz-piotr) [reply waffle-compiler/src/config/inputToConfig.ts:19]: I believe the configuration could totally work with the following keys and defaults:
```js
{
  "outputDirectory": "./build",
  "flattenerOutputDirectory": "./build/flat"
}
```

File: .gitignore [hunk 18]
```diff
@@ -3,3 +3,4 @@ dist
 yarn-error.log
 /docs/build
 /docs/env
+.idea
```
Comment (marekkirejczyk) [inline .gitignore:6]: extract to a separate PR

File: waffle-compiler/test/flattener/flattener.ts [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+  await flattenAndSave(config);
+  return fs.readFileSync('./test/flattener/flattenFiles/child.sol', 'utf8');
+};
+
+describe('flattening', () => {
+  after(async () => {
+    await fs.remove(flattenOutputDirectory);
+  });
+
+  it('should include the parent in the child file', async () => {
```
Comment (marekkirejczyk) [inline waffle-compiler/test/flattener/flattener.ts:25]: I would add end-to-end test, perhaps sth like:
```ts
const flattenResultPath = './test/flattener/flattenFiles/result.sol'
...
it('flattens files', async () => {
  const file = await flatAndGetChild();
  await fs.remove(flattenResultPath));
});
```

File: waffle-compiler/test/flattener/flattenFiles/parent.sol [hunk 20]
```diff
@@ -0,0 +1,75 @@
+// Dependency file: test/flattener/testSource/cycle1.sol
```
Comment (marekkirejczyk) [inline waffle-compiler/test/flattener/flattenFiles/parent.sol:1]: Shouldn't `waffle-compiler/test/flattener/flattenFiles/` be in the .gitignore? 
It seems to be temporary test output.

File: waffle-compiler/test/flattener/flattener.ts [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+});
+
+const flatAndGetChild = async () => {
+  await flattenAndSave(config);
+  return fs.readFileSync('./test/flattener/flattenFiles/child.sol', 'utf8');
+};
+
+describe('flattening', () => {
+  after(async () => {
+    await fs.remove(flattenOutputDirectory);
```
Comment (marekkirejczyk) [inline waffle-compiler/test/flattener/flattener.ts:22]: This is very problematic together with #tmpoutput, as after test run, you will have changes to stage in git.
Remove `waffle-compiler/test/flattener/flattenFiles` from repo and add it to `.gitignore` to fix it.

File: waffle-compiler/.gitignore [hunk 22]
```diff
@@ -1,2 +1,3 @@
 /**/custom_build
 /**/build
+/test/flattener/flattenFiles
```
Comment (marekkirejczyk) [inline waffle-compiler/.gitignore:3]: Files in `waffle-compiler/test/flattener/*` are still a part of this PR

File: waffle-compiler/src/flattener.ts [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+    let source = '';
+
+    contract.map(dependency => {
+      source = modifyDependency(dependency, contract, source);
+    });
+    fileSystem.writeFile(filePath, source);
+  });
+}
+
+function modifyDependency(
```
Comment (marekkirejczyk) [inline waffle-compiler/src/flattener.ts:67]: `replaceDirectivesWithComments`
